File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/virology/tokengen/VirologyProcessorService.java
package uk.nhs.nhsx.virology.tokengen;

import uk.nhs.nhsx.core.events.Events;
import uk.nhs.nhsx.core.events.ExceptionThrown;
import uk.nhs.nhsx.virology.CtaToken;
import uk.nhs.nhsx.virology.TestKit;
import uk.nhs.nhsx.virology.VirologyService;
import uk.nhs.nhsx.virology.result.VirologyTokenGenRequest;
import uk.nhs.nhsx.virology.result.VirologyTokenGenResponse;

import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import static java.util.concurrent.CompletableFuture.allOf;
import static java.util.concurrent.CompletableFuture.supplyAsync;
import static java.util.stream.IntStream.range;

public class VirologyProcessorService {

    private final VirologyService virologyService;
    private final VirologyProcessorStore virologyProcessorStore;
    private final Supplier<Instant> clock;
    private final int maxRetryCount;
    private final Events events;

    private final ExecutorService executor = Executors.newFixedThreadPool(20);

    public VirologyProcessorService(VirologyService virologyService,
                                    VirologyProcessorStore virologyProcessorStore,
                                    Supplier<Instant> clock,
                                    int maxRetryCount,
                                    Events events) {
        this.virologyService = virologyService;
        this.virologyProcessorStore = virologyProcessorStore;
        this.clock = clock;
        this.maxRetryCount = maxRetryCount;
        this.events = events;
    }

    public CtaProcessorResult generateAndStoreTokens(CtaProcessorRequest event) {
        var start = clock.get();
        var tokens = generateTokens(event);

        if (tokens.isEmpty()) {
            return new CtaProcessorResult.Error("No tokens generated/stored (empty list)");
        }

        var filename = start.toString();
        var ctaTokensZip = storeTokens(event, tokens, filename);

        var message = String.format(
            "Generated %d tokens in %d seconds",
            tokens.size(), Duration.between(start, Instant.now()).toSeconds()
        );

        return new CtaProcessorResult.Success(ctaTokensZip.filename, message);
    }

    private CtaTokensZip storeTokens(CtaProcessorRequest event, List<CtaToken> tokens, String filename) {
        var csvContent = VirologyProcessorExports.csvFrom(tokens, event.testResult, event.testEndDate);
        var ctaTokensCsv = new CtaTokensCsv(filename, csvContent);
        virologyProcessorStore.storeCsv(ctaTokensCsv);

        var zipFile = VirologyProcessorExports.zipFrom(filename, ctaTokensCsv);
        var ctaTokensZip = new CtaTokensZip(filename, zipFile);
        virologyProcessorStore.storeZip(ctaTokensZip);
        return ctaTokensZip;
    }

    private List<CtaToken> generateTokens(CtaProcessorRequest event) {
        var request = new VirologyTokenGenRequest(event.testResult, event.testEndDate, TestKit.valueOf(event.testKit));
        var futures = createTokenGenFutures(event, request);
        return collectResultsFrom(futures);
    }

    private List<CompletableFuture<VirologyTokenGenResponse>> createTokenGenFutures(CtaProcessorRequest event,
                                                                                    VirologyTokenGenRequest request) {
        return range(0, event.numberOfTokens)
            .mapToObj(it ->
                supplyAsync(() -> generateTokenRetryingOnFailure(request), executor)
                    .exceptionally(ex -> {
                        events.emit(getClass(), new ExceptionThrown<>(ex, "Exception while generating CTA token"));
                        return null;
                    })
            )
            .collect(Collectors.toList());
    }

    private List<CtaToken> collectResultsFrom(List<CompletableFuture<VirologyTokenGenResponse>> futures) {
        return allOf(futures.toArray(new CompletableFuture[0]))
            .thenApply(
                future -> futures
                    .stream()
                    .map(CompletableFuture::join)
                    .filter(Objects::nonNull)
                    .map(it -> CtaToken.of(it.ctaToken))
                    .collect(Collectors.toList())
            )
            .join();
    }

    private VirologyTokenGenResponse generateTokenRetryingOnFailure(VirologyTokenGenRequest request) {
        var numberOfTries = 0;
        do {
            try {
                return virologyService.acceptTestResultGeneratingTokens(request);
            } catch (Exception e) {
                numberOfTries++;
            }
        } while (numberOfTries < maxRetryCount);

        throw new RuntimeException("Generate cta token exceeded maximum of " + maxRetryCount + " retries");
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/virology/tokengen/VirologyProcessorService.kt
package uk.nhs.nhsx.virology.tokengen

import uk.nhs.nhsx.core.events.Events
import uk.nhs.nhsx.core.events.ExceptionThrown
import uk.nhs.nhsx.virology.CtaToken
import uk.nhs.nhsx.virology.CtaTokensAndUrlGenerationFailed
import uk.nhs.nhsx.virology.TestKit
import uk.nhs.nhsx.virology.VirologyService
import uk.nhs.nhsx.virology.result.TestEndDate
import uk.nhs.nhsx.virology.result.TestResult
import uk.nhs.nhsx.virology.result.VirologyTokenGenRequestV2
import uk.nhs.nhsx.virology.result.VirologyTokenGenResponse
import uk.nhs.nhsx.virology.tokengen.CtaProcessorResult.Error
import uk.nhs.nhsx.virology.tokengen.CtaProcessorResult.Success
import java.time.Duration
import java.time.Instant
import java.time.ZonedDateTime
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import java.util.function.Supplier
import java.util.stream.Collectors
import java.util.stream.IntStream

class VirologyProcessorService(
    private val virologyService: VirologyService,
    private val virologyProcessorStore: VirologyProcessorStore,
    private val clock: Supplier<Instant>,
    private val virologyProcessorExports: VirologyProcessorExports,
    private val maxRetryCount: Int,
    private val events: Events
) {
    private val executor = Executors.newFixedThreadPool(20)

    fun generateAndStoreTokens(event: CtaProcessorRequest): CtaProcessorResult {
        val start = clock.get()
        val tokens = generateTokens(event)
        if (tokens.isEmpty()) {
            return Error("No tokens generated/stored (empty list)")
        }
        val filename = start.toString().replace(":","")
        val ctaTokensZip = storeTokens(event, tokens, filename)
        val message = String.format("Generated %d tokens in %d seconds", tokens.size, Duration.between(start, Instant.now()).toSeconds())
        return Success(ctaTokensZip.filename, message)
    }

    fun generateStoreTokensAndReturnSignedUrl(requests: List<CtaTokenZipFileEntryRequest>, zipFilePassword: String,linkExpirationDate: ZonedDateTime): CtaTokenZipFile {
        val tokenFile = generateAndStoreTokensAsSingleZip(requests, zipFilePassword)
        if (tokenFile !is Success) {
            events.emit(javaClass, CtaTokensAndUrlGenerationFailed(requests, "Token file generation failed"))
            throw RuntimeException("Token file generation failed")
        }
        val filename = tokenFile.zipFilename
        val url = virologyProcessorStore.generateSignedURL(filename, Date.from(linkExpirationDate.toInstant()))

        if(url == null) {
            events.emit(javaClass,CtaTokensAndUrlGenerationFailed(requests, "Unable to retrieve signed URL from S3"))
            throw RuntimeException("Unable to retrieve signed URL from S3")
        }

        return CtaTokenZipFile(url, filename, zipFilePassword)
    }

    private fun generateAndStoreTokensAsSingleZip(requests: List<CtaTokenZipFileEntryRequest>, zipFilePassword: String): CtaProcessorResult {
        val start = clock.get()
        val zipFileEntries = requests.stream().map { (testResult, testEndDate, testKit, filename, numberOfTokens) ->
            CtaTokenZipFileEntry(
                testResult,
                testEndDate,
                filename,
                generateTokens(
                    CtaProcessorRequest(
                        testResult,
                        testEndDate,
                        testKit,
                        numberOfTokens
                    )
                )
            )
        }.collect(Collectors.toList())
        val filename = start.toString().replace(":", "")
        val ctaTokensZip = storeTokensAsSingleZip(zipFileEntries, filename, zipFilePassword)
        val message = String.format(
            "Generated %d token entries in %d seconds",
            zipFileEntries.size, Duration.between(start, Instant.now()).toSeconds()
        )
        return Success(ctaTokensZip.filename, message)
    }

    private fun storeTokensAsSingleZip(zipFileEntries: List<CtaTokenZipFileEntry>, zipFileName: String, zipFilePassword: String): CtaTokensZip {
        val ctaTokenCsvs = zipFileEntries
            .map { (testResult, endDate, filename, ctaTokens) ->
                val csvContent =
                    virologyProcessorExports.csvFrom(ctaTokens, testResult, endDate)
                CtaTokensCsv(filename, csvContent)
            }
        val zipFile = virologyProcessorExports.zipFrom(zipFileName, ctaTokenCsvs, zipFilePassword)
        val ctaTokensZip = CtaTokensZip(zipFileName, zipFile)
        virologyProcessorStore.storeZip(ctaTokensZip)
        return ctaTokensZip
    }

    private fun storeTokens(event: CtaProcessorRequest, tokens: List<CtaToken>, filename: String): CtaTokensZip {
        val csvContent = virologyProcessorExports.csvFrom(tokens, event.testResult, event.testEndDate)
        val ctaTokensCsv = CtaTokensCsv(filename, csvContent)
        virologyProcessorStore.storeCsv(ctaTokensCsv)
        val zipFile = virologyProcessorExports.zipFrom(filename, listOf(ctaTokensCsv))
        val ctaTokensZip = CtaTokensZip(filename, zipFile)
        virologyProcessorStore.storeZip(ctaTokensZip)
        return ctaTokensZip
    }

    private fun generateTokens(event: CtaProcessorRequest) = collectResultsFrom(
        createTokenGenFutures(
            event,
            VirologyTokenGenRequestV2(
                event.testEndDate,
                event.testResult,
                event.testKit
            )
        )
    )

    private fun createTokenGenFutures(
        event: CtaProcessorRequest,
        request: VirologyTokenGenRequestV2
    ) = IntStream.range(0, event.numberOfTokens)
        .mapToObj {
            CompletableFuture.supplyAsync({ generateTokenRetryingOnFailure(request) }, executor)
                .exceptionally { ex: Throwable ->
                    events.emit(javaClass, ExceptionThrown(ex, "Exception while generating CTA token"))
                    null
                }
        }
        .collect(Collectors.toList())

    private fun collectResultsFrom(futures: List<CompletableFuture<VirologyTokenGenResponse>>) =
        CompletableFuture.allOf(*futures.toTypedArray<CompletableFuture<*>>())
            .thenApply {
                futures
                    .stream()
                    .map { obj: CompletableFuture<VirologyTokenGenResponse> -> obj.join() }
                    .filter { obj: VirologyTokenGenResponse? -> Objects.nonNull(obj) }
                    .map(VirologyTokenGenResponse::ctaToken)
                    .collect(Collectors.toList())
            }
            .join()

    private fun generateTokenRetryingOnFailure(request: VirologyTokenGenRequestV2): VirologyTokenGenResponse {
        var numberOfTries = 0
        do {
            try {
                return virologyService.acceptTestResultGeneratingTokens(request)
            } catch (e: Exception) {
                numberOfTries++
            }
        } while (numberOfTries < maxRetryCount)
        throw RuntimeException("Generate cta token exceeded maximum of $maxRetryCount retries")
    }
}

data class CtaTokenZipFileEntry(val testResult: TestResult, val endDate: TestEndDate, val filename: String, val ctaTokens: List<CtaToken>)
data class CtaTokenZipFileEntryRequest (val testResult: TestResult, val testEndDate:TestEndDate, val testKit:TestKit, val filename:String, val numberOfTokens:Int)

